Sources/NIOWebSocket/NIOWebSocketServerUpgrader.swift

```diff
@@ -224,62 +224,177 @@ public final class NIOWebSocketServerUpgrader: HTTPServerProtocolUpgrader, @unch
+            upgradeRequest: upgradeRequest,
+            maxFrameSize: self.maxFrameSize,
+            automaticErrorHandling: self.automaticErrorHandling,
+            upgradePipelineHandler: self.upgradePipelineHandler
+        )
+    }
+}
+
+private func _buildUpgradeResponse(
+    channel: Channel,
+    upgradeRequest: HTTPRequestHead,
+    initialResponseHeaders: HTTPHeaders,
+    shouldUpgrade: @Sendable (Channel, HTTPRequestHead) -> EventLoopFuture<HTTPHeaders?>
+) -> EventLoopFuture<HTTPHeaders> {
+    let key: String
+    let version: String
+
+    do {
+        key = try upgradeRequest.headers.nonListHeader("Sec-WebSocket-Key")
+        version = try upgradeRequest.headers.nonListHeader("Sec-WebSocket-Version")
+    } catch {
+        return channel.eventLoop.makeFailedFuture(error)
+    }
+
+    // The version must be 13.
+    guard version == "13" else {
+        return channel.eventLoop.makeFailedFuture(NIOWebSocketUpgradeError.invalidUpgradeHeader)
+    }
+
+    return shouldUpgrade(channel, upgradeRequest).flatMapThrowing { extraHeaders in
+        guard let extraHeaders = extraHeaders else {
+            throw NIOWebSocketUpgradeError.unsupportedWebSocketTarget
         }
+        return extraHeaders
+    }.map { (extraHeaders: HTTPHeaders) in
+        var extraHeaders = extraHeaders
 
-        if self.automaticErrorHandling {
-            upgradeFuture = upgradeFuture.flatMap {
-                context.pipeline.addHandler(WebSocketProtocolErrorHandler())
-            }
+        // Cool, we're good to go! Let's do our upgrade. We do this by concatenating the magic
+        // GUID to the base64-encoded key and taking a SHA1 hash of the result.
+        let acceptValue: String
+        do {
+            var hasher = SHA1()
+            hasher.update(string: key)
+            hasher.update(string: magicWebSocketGUID)
+            acceptValue = String(base64Encoding: hasher.finish())
         }
 
-        return upgradeFuture.flatMap {
-            self.upgradePipelineHandler(context.channel, upgradeRequest)
+        extraHeaders.replaceOrAdd(name: "Upgrade", value: "websocket")
+        extraHeaders.add(name: "Sec-WebSocket-Accept", value: acceptValue)
+        extraHeaders.replaceOrAdd(name: "Connection", value: "upgrade")
+
+        return extraHeaders
+    }
+}
+
+private func _upgrade<UpgradeResult>(
+    channel: Channel,
+    upgradeRequest: HTTPRequestHead,
+    maxFrameSize: Int,
+    automaticErrorHandling: Bool,
+    upgradePipelineHandler: @escaping @Sendable (Channel, HTTPRequestHead) -> EventLoopFuture<UpgradeResult>
+) -> EventLoopFuture<UpgradeResult> {
+    /// We never use the automatic error handling feature of the WebSocketFrameDecoder: we always use the separate channel
+    /// handler.
+    var upgradeFuture = channel.pipeline.addHandler(WebSocketFrameEncoder()).flatMap {
+        channel.pipeline.addHandler(ByteToMessageHandler(WebSocketFrameDecoder(maxFrameSize: maxFrameSize)))
+    }
```

Comment:
can we `addHandlers` here instead?

Reply:
or even have a sync version for adding all these handlers and then do the EL dance